### PR TITLE
300 table generation

### LINKED ISSUE #300: Improve table generation
Change the table generation to use pandas to convert numpy arrays of data into rst and html tables, instead of generating these by hand.

This might allow us to remove docutils as a dependency.

### PR DESCRIPTION
**Pull request by wathen**
#### Description of Work

Fixes #300 

Converts hard coded tables into pandas tables.

#### Testing Instructions

1. Check tables give the same results as before.
2. Check tables have right colour scales.

Function: Does the change do what it's supposed to?

Tests: Does it pass? Is there adequate coverage for new code?

Style: Is the coding style consistent? Is anything overly confusing?

Documentation: Is there a suitable change to documentation for this change?

### CODE REVIEW

**Comment by AndrewLister-STFC** [comment]
Can you also update pictures in the docs and readme files

File: fitbenchmarking/results_output.py [hunk 1]
```diff
@@ -3,12 +3,16 @@
 """
 
 from __future__ import (absolute_import, division, print_function)
-
+from collections import OrderedDict
 import logging
+import os
+import pandas as pd
+import numpy as np
```
[the remark this answers is not in the document]
Comment (wathen) [reply fitbenchmarking/results_output.py]: Done

File: fitbenchmarking/fitbenchmarking_default_options.json [hunk 2]
```diff
@@ -7,13 +7,13 @@
...
+        "sasview" : ["amoeba", "lm-bumps", "newton", "de", "pt", "mp"],
         "dfogn" : ["dfogn"],
         "ralfit" : ["gn","gn_reg","hybrid","hybrid_reg"],
 	"minuit" : ["minuit"]
         },
 
     "comparison_mode" : "both",
 
-    "num_runs" : 5
+    "num_runs" : 1
```
[the remark this answers is not in the document]
Comment (wathen) [reply fitbenchmarking/fitbenchmarking_default_options.json]: Done

File: fitbenchmarking/results_output.py [hunk 3]
```diff
@@ -18,21 +22,28 @@
...
-    @param software_options :: dictionary containing software used in fitting the problem, list of minimizers and location of json file contain minimizers
-    @param minimizers :: array with minimizer names
-    @param results_per_test :: results nested array of objects
-    @param group_name :: name of the problem group
-    @param use_errors :: bool whether to use errors or not
-    @param color_scale :: color the html table
-    @param results_dir :: directory in which the results are saved
-
-    @returns :: html/rst tables with the fitting results
+    :param software_options :: dictionary containing software used in fitting the problem, list of minimizers and location of json file contain minimizers
```
[the remark this answers is not in the document]
Comment (wathen) [reply fitbenchmarking/results_output.py]: Might as well do it now.

File: example_scripts/example_runScripts_expert.py [hunk 4]
```diff
@@ -116,29 +115,38 @@ def main(argv):
...
+                comparison_mode = 'both'
 
-        # Saves runtime minimizer results
-        save_tables(tables_dir, runtime_tbl, use_errors, group_name, 'runtime')
+        tables_dir = create_dirs.restables_dir(results_dir, group_name)
+        linked_problems = \
+            visual_pages.create_linked_probs(prob_results, group_name, results_dir)
+
+        table_name = []
+        weighted_values = {True: 'weighted', False: 'unweighted'}
```
[the remark this answers is not in the document]
Comment (wathen) [reply example_scripts/example_runScripts_expert.py]: Change this in `results_output.py` on line 66

File: fitbenchmarking/results_output.py [hunk 5]
```diff
@@ -52,140 +63,236 @@ def save_results_tables(software_options, results_per_test, group_name,
...
+    :param acc_tbl :: DataFrame of the accuracy results
+    :type acc_tbl :: pandas DataFrame
+    :param runtime_tbl :: DataFrame of the timing results
+    :type runtime_tbl :: pandas DataFrame
+    :param minimizers :: list of minimizers (column headers)
+    :type minimizers :: list
+    :param colour_scale :: user defined colour scale
+    :type colour_scale :: list
+    :param html_links :: html links used in pandas rendering
+    :type html_links :: list
```
[the remark this answers is not in the document]
Comment (wathen) [reply fitbenchmarking/results_output.py]: Done

File: fitbenchmarking/results_output.py [hunk 6]
```diff
@@ -18,21 +22,28 @@
 FILENAME_EXT_TXT = 'txt'
 FILENAME_EXT_HTML = 'html'
 
+html_color_scale = ['#fef0d9', '#fdcc8a', '#fc8d59', '#e34a33', '#b30000']
```
[the remark this answers is not in the document]
Comment (wathen) [reply fitbenchmarking/results_output.py]: Done

File: fitbenchmarking/results_output.py [hunk 7]
```diff
@@ -52,140 +63,236 @@ def save_results_tables(software_options, results_per_test, group_name,
...
+    :param acc_tbl :: DataFrame of the accuracy results
+    :type acc_tbl :: pandas DataFrame
+    :param runtime_tbl :: DataFrame of the timing results
+    :type runtime_tbl :: pandas DataFrame
+    :param minimizers :: list of minimizers (column headers)
+    :type minimizers :: list
+    :param colour_scale :: user defined colour scale
+    :type colour_scale :: list
+    :param rst_links :: rst links used in pandas rendering
+    :type rst_links :: list
```
[the remark this answers is not in the document]
Comment (wathen) [reply fitbenchmarking/results_output.py]: Done

File: fitbenchmarking/results_output.py [hunk 8]
```diff
@@ -52,140 +63,236 @@ def save_results_tables(software_options, results_per_test, group_name,
...
+    return acc_results, time_results, html_links
 
 
-def generate_tables(results_per_test, minimizers):
+def create_pandas_dataframe(table_data, minimizers):
     """
-    Generates accuracy and runtime tables, with both normalised and absolute results, and summary tables.
+    Generates pandas tables.
+
+    :param table_data :: dictionary containing results
```
[the remark this answers is not in the document]
Comment (wathen) [reply fitbenchmarking/results_output.py]: Added more info in docstring

File: fitbenchmarking/fitbenchmarking_default_options.json [hunk 9]
```diff
@@ -7,13 +7,13 @@
                     "Simplex","SteepestDescent",
                     "Trust Region"],
         "scipy" : ["lm", "trf", "dogbox"],
-        "sasview" : ["amoeba", "lm", "newton", "de", "pt", "mp"],
+        "sasview" : ["amoeba", "lm-bumps", "newton", "de", "pt", "mp"],
```
[the remark this answers is not in the document]
Comment (wathen) [reply fitbenchmarking/fitbenchmarking_default_options.json]: Yea I agree

File: fitbenchmarking/results_output.py [hunk 10]
```diff
@@ -52,140 +63,236 @@ def save_results_tables(software_options, results_per_test, group_name,
     if isinstance(software, list):
         minimizers = sum(minimizers, [])
 
+    weighted_values = {True: 'weighted', False: 'unweighted'}
```
[the remark this answers is not in the document]
Comment (wathen) [reply fitbenchmarking/results_output.py]: See above

File: fitbenchmarking/results_output.py [hunk 11]
```diff
@@ -52,140 +63,236 @@ def save_results_tables(software_options, results_per_test, group_name,
...
+    :type results_per_test :: list[list[list]]
+    :param minimizers :: array with minimizer names
+    :type minimizers :: list
+    :param linked_problems :: rst links for supporting pages
+    :type linked_problems :: list[str]
+    :param colour_scale :: colour the html table
+    :type colour_scale :: list
+    :param comparison_mode :: check whether to produced 'rel', 'abs' or 'both'
+                              tables
+    :type comparison_mode :: str
```
[the remark this answers is not in the document]
Comment (wathen) [reply fitbenchmarking/results_output.py]: done

File: fitbenchmarking/results_output.py [hunk 12]
```diff
@@ -52,140 +63,236 @@ def save_results_tables(software_options, results_per_test, group_name,
...
 
-    return tbl_runtime_indiv
+    count = 1
+    prev_name = ''
+    template = '<a target="_blank" href="{0}">{1}</a>'
+    acc_results = OrderedDict()
+    time_results = OrderedDict()
+    html_links = []
+
+    for test_idx, prob_results in enumerate(results_per_test):
```
[the remark this answers is not in the document]
Comment (wathen) [reply fitbenchmarking/results_output.py]: Done

File: fitbenchmarking/results_output.py [hunk 13]
```diff
@@ -52,140 +63,236 @@ def save_results_tables(software_options, results_per_test, group_name,
...
-    return accuracy_tbl, runtime_tbl, sum_cells_acc, sum_cells_runtime
+def check_normalised(data, colours, colour_bounds):
+    """
+    Loops through row data of pandas data frame and assigns
+    colours depending on size.
 
+    :param data :: Pandas row data
+    :type data :: Pandas Series
+    :param colours :: rst or html colour definitions
+    :type colours :: list[str]
```
[the remark this answers is not in the document]
Comment (wathen) [reply fitbenchmarking/results_output.py]: Done

File: fitbenchmarking/results_output.py [hunk 14]
```diff
@@ -52,140 +63,236 @@ def save_results_tables(software_options, results_per_test, group_name,
...
-                                   color_scale=color_scale,
-                                   comparison_mode=comparison_mode)
-
-    save_tables(tables_dir, acc_tbl, use_errors, group_name,
-                FILENAME_SUFFIX_ACCURACY)
-    save_tables(tables_dir, runtime_tbl, use_errors, group_name,
-                FILENAME_SUFFIX_RUNTIME)
-
-    # Shut down logging at end of run
+    table_name = []
```
[the remark this answers is not in the document]
Comment (wathen) [reply fitbenchmarking/results_output.py]: Done

File: fitbenchmarking/results_output.py [hunk 15]
```diff
@@ -52,140 +63,236 @@ def save_results_tables(software_options, results_per_test, group_name,
...
-                                     comparison_type='accuracy',
-                                     comparison_dim='',
-                                     using_errors=use_errors,
-                                     color_scale=color_scale,
-                                     comparison_mode=comparison_mode)
 
-    return tbl_acc_indiv
+    acc_dict, time_dict, html_links = create_results_dict(results_per_test,
+                                                          linked_problems)
+    acc_tbl = \
```
[the remark this answers is not in the document]
Comment (wathen) [reply fitbenchmarking/results_output.py]: Done

File: fitbenchmarking/results_output.py [hunk 16]
```diff
@@ -18,21 +22,28 @@
 FILENAME_EXT_TXT = 'txt'
 FILENAME_EXT_HTML = 'html'
 
+HTML_COLUR_SCALE = ['#fef0d9', '#fdcc8a', '#fc8d59', '#e34a33', '#b30000']
```
Comment (AndrewLister-STFC) [inline fitbenchmarking/results_output.py]: Typo in COLOUR (or COLOR if you were going for the spelling that was there before)

**Pull request merged**